kernel/include/kernel_internal.h

```diff
@@ -51,30 +52,61 @@ extern void _setup_new_thread(struct k_thread *new_thread,
 			      void *p1, void *p2, void *p3,
 			      int prio, u32_t options);
 
-/* context switching and scheduling-related routines */
-
-extern unsigned int __swap(unsigned int key);
-
 #ifdef CONFIG_TIMESLICING
 extern void _update_time_slice_before_swap(void);
+#else
+#define _update_time_slice_before_swap() /**/
 #endif
 
 #ifdef CONFIG_STACK_SENTINEL
 extern void _check_stack_sentinel(void);
+#else
+#define _check_stack_sentinel() /**/
 #endif
 
+/* context switching and scheduling-related routines */
+#ifdef CONFIG_USE_SWITCH
+
+/* New style context switching.  _arch_switch() is a lower level
+ * primitive that doesn't know about the scheduler or return value.
+ * Needed for SMP, where the scheduler requires spinlocking that we
+ * don't want to have to do in per-architecture assembly.
+ */
 static inline unsigned int _Swap(unsigned int key)
 {
+	struct k_thread *new_thread, *old_thread;
+
+	old_thread = _kernel.current;
 
-#ifdef CONFIG_STACK_SENTINEL
 	_check_stack_sentinel();
```

Comment:
Me personally, I always preferred explicit ifdefs rather than defining various functions to empty macros, as I think it makes it clearer that this call site doesn't do anything.

I used to lose arguments with Dirk all the time about this.

Carry on, I'm just whining.